runtime/src/bank.rs

```diff
@@ -1455,11 +1455,69 @@ impl Bank {
             }
         });
 
+        let (_, recompilation_time_us) = measure_us!({
+            // Recompile loaded programs one at a time before the next epoch hits
+            let (_epoch, slot_index) = new.get_epoch_and_slot_index(new.slot());
+            let mut loaded_programs_cache = new.loaded_programs_cache.write().unwrap();
+            if loaded_programs_cache.upcoming_environments.is_some() {
+                if let Some((key, program_to_recompile)) =
+                    loaded_programs_cache.programs_to_recompile.pop()
+                {
+                    drop(loaded_programs_cache);
+                    let recompiled = new.load_program(&key, false, Some(program_to_recompile));
+                    let mut loaded_programs_cache = new.loaded_programs_cache.write().unwrap();
+                    loaded_programs_cache.replenish(key, recompiled);
+                }
+            } else if new.epoch() != loaded_programs_cache.latest_root_epoch
```

Comment:
Could a malicious leader trigger this check by adding a bank in a future epoch? Does it make sense to make this check more specific (e.g. new.epoch == loaded_programs_cache.latest_root_epoch + 1, and slot is within N slots in the new epoch)?

Reply:
Aren't leaders bound by the leader schedule in that they can't just submit a block for any slot, but only the one they were assigned to?

Reply:
https://github.com/solana-labs/solana/blob/1814b2bc81aca29b322f8e1f280494c1b9ee6b32/turbine/src/sigverify_shreds.rs#L141C6-L141C6

> Would making this check more strict cause any problems?

Probably not, but haven't thought it through.